src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -561,6 +561,110 @@ FCIMPL2(Object*, RuntimeTypeHandle::CreateInstanceForGenericType, ReflectClassBa
+ * a boxed zero-inited instance of the value type.
+ */
+    void QCALLTYPE RuntimeTypeHandle::GetNewobjHelperFnPtr(
+        QCall::TypeHandle pTypeHandle,
+        PCODE* ppNewobjHelper,
+        MethodTable** ppMT,
+        BOOL fUnwrapNullable,
+        BOOL fAllowCom)
+{
+    CONTRACTL{
+        QCALL_CHECK;
+        PRECONDITION(CheckPointer(ppNewobjHelper));
+        PRECONDITION(CheckPointer(ppMT));
+        PRECONDITION(*ppNewobjHelper == NULL);
+        PRECONDITION(*ppMT == NULL);
+    }
+    CONTRACTL_END;
+
+    BEGIN_QCALL;
+
+    TypeHandle typeHandle = pTypeHandle.AsTypeHandle();
+
+    // Don't allow arrays, pointers, byrefs, or function pointers.
+    if (typeHandle.IsTypeDesc() || typeHandle.IsArray())
+    {
+        COMPlusThrow(kArgumentException, W("Argument_InvalidValue"));
+    }
+
+    MethodTable* pMT = typeHandle.AsMethodTable();
+    PREFIX_ASSUME(pMT != NULL);
+
+    // Don't allow creating instances of void or delegates
+    if (pMT == MscorlibBinder::GetElementType(ELEMENT_TYPE_VOID) || pMT->IsDelegate())
+    {
+        COMPlusThrow(kArgumentException, W("Argument_InvalidValue"));
+    }
+
+    // Don't allow string or string-like (variable length) types.
+    if (pMT->HasComponentSize())
+    {
+        COMPlusThrow(kArgumentException, W("Argument_NoUninitializedStrings"));
+    }
+
+    // Don't allow abstract classes or interface types
+    if (pMT->IsAbstract()) {
+        COMPlusThrow(kMemberAccessException, W("Acc_CreateAbst"));
+    }
+
+    // Don't allow open generics or generics instantiated over __Canon
+    if (pMT->ContainsGenericVariables()) {
+        COMPlusThrow(kMemberAccessException, W("Acc_CreateGeneric"));
+    }
+    if (pMT->IsSharedByGenericInstantiations()) {
+        COMPlusThrow(kNotSupportedException, W("NotSupported_Type"));
+    }
+
+    // Don't allow ref structs
+    if (pMT->IsByRefLike()) {
+        COMPlusThrow(kNotSupportedException, W("NotSupported_ByRefLike"));
+    }
+
+    // Never allow the allocation of an unitialized ContextBoundObject derived type, these must always be created with a paired
+    // transparent proxy or the jit will get confused.
+
+#ifdef FEATURE_COMINTEROP
+    // Unless caller allows, do not allow allocation of uninitialized RCWs (COM objects).
+    // If the caller allows this, getNewHelperStatic will return an appropriate allocator.
+    if (!fAllowCom && pMT->IsComObjectType())
+    {
+        COMPlusThrow(kNotSupportedException, W("NotSupported_ManagedActivation"));
+    }
+#endif // FEATURE_COMINTEROP
+
+    // If the caller passed Nullable<T> but asked us to unwrap nullable types,
+    // instead pretend they had passed the 'T' directly.
+    if (fUnwrapNullable && Nullable::IsNullableType(pMT))
+    {
+        pMT = pMT->GetInstantiation()[0].GetMethodTable();
+    }
+
+    // Ensure the type's cctor has run
+    Assembly* pAssem = pMT->GetAssembly();
+    if (!pMT->IsClassInited())
```

Comment:
I think this will change the semantics of GetUnitializedObject slightly. AFAIK GetUnitializedObject triggers the cctor based on whether the type is `beforefieldinit` or not. Not sure how much we care.

Reply:
This logic was copied from `RunClassConstructor`. Though in hindsight and per your comments, it's probably best that I instead duplicate the logic that's in `CreateInstance` and `Allocate`. That would be something like:

```cpp
Assembly* pAssem = pMT->GetAssembly(); // not used, but CreateInstance makes the same call - perhaps for a side effect?
pMT->EnsureInstanceActive(); // called unconditionally by MethodTable::Allocate
if (pMT->HasPreciseInitCctors())
{
    pMT->CheckRunClassInitAsIfConstructingThrowing();
}
```